src/LaunchDarkly.EventSource/ByteArrayLineScanner.cs

```diff
@@ -0,0 +1,157 @@
+                    (int)_partialLine.Position - 1); // don't include the CR
+                _partialLine = null;
+                return true;
+            }
+
+            int startedAt = _startPos, pos = _startPos;
+
+            while (pos < _count)
+            {
+                var b = _buffer[pos];
+                if (b == '\n') // LF by itself terminates a line
+                {
+                    _startPos = pos + 1; // next line will start after the LF
+                    break;
+                }
+                if (b == '\r')
+                {
+                    if (pos < (_count - 1))
+                    {
+                        _startPos = pos + 1; // next line will start after the CR--
+                        if (_buffer[pos + 1] == '\n') // --unless there was an LF right after that
+                        {
+                            _startPos++;
+                        }
+                        break;
+                    }
+                    else
+                    {
+                        // CR by itself and CR+LF are both valid line endings in SSE, so if the very
+                        // last character we saw was a CR, we can't know when the line is fully read
+                        // until we've gotten more data. So we'll need to treat this as an incomplete
+                        // line.
+                        pos++;
+                        break;
+                    }
+                }
+                pos++;
+            }
+
+            if (pos == _count) // we didn't find a line terminator
+            {
+                lineOut = new Utf8ByteSpan();
+                if (_count < _capacity)
+                {
+                    // There's still room in the buffer, so we'll re-scan the line once they add more bytes
+                    return false;
+                }
+                // We need to dump the incomplete line into _partialLine so we can make room in the buffer
+                var partialCount = pos - _startPos;
+                if (_partialLine is null)
+                {
+                    _partialLine = new MemoryStream(partialCount);
+                }
+                _partialLine.Write(_buffer, _startPos, partialCount);
+
+                // Clear the main buffer
+                _startPos = _count = 0;
+
+                return false;
+            }
+
+            if (_partialLine != null && _partialLine.Position > 0)
+            {
+                _partialLine.Write(_buffer, startedAt, pos - startedAt);
+                lineOut = new Utf8ByteSpan(_partialLine.GetBuffer(), 0, (int)_partialLine.Position);
+                _partialLine = null;
+
+                // If there are still bytes in the buffer, move them over to make more room
+                if (_startPos < _count)
+                {
+                    System.Buffer.BlockCopy(_buffer, _startPos, _buffer, 0, _count - _startPos);
```

Comment:
Is there a thought process here for doing this shift copy in the case where there was a `_partialLine` vs the reset when the buffer bytes ended on a line ending?

Thinking about it more, I would guess that's because we normally expect from the SSE stream. I guess it makes sense to me then, but took a fair bit of thought.

Reply:
I'm not sure I understand the question; can you be more specific about which code path you mean by "the reset when the buffer bytes ended on a line ending"? If you mean line 151 - the answer is in the question. That is, if we have no more bytes in the buffer, then there's nothing to shift over.

Reply:
This comment ended up being more stream of consciousness than a true question, so I would consider it resolved anyway. I'll clarify what I was originally unsure about however.

My question was more why we only check `_startPos == _count`, on line 149, rather than doing a `_startPos < _count` and performing this same `BlockCopy` to shift the bytes and make room for more input. In the case that the provided bytes terminate in a line ending, the check on 149 is sufficient, but will leave already consumed input taking space in the buffer in other cases.

My confusion was just about why we would bother with the `BlockCopy` in one case but not the other, but I believe the statement after my previous question explains the difference by considering that the network data being given to the SSE stream will normally be line-chunked anyway.

Reply:
No, the assumption was _not_ that the stream will normally be line-chunked. That may be true, but it's definitely not why I wrote the code this way. Let me take another look and see if I can understand better what you're getting at.

Reply:
Oh wait, I see what you're saying. There is a different rationale, but it's not self-evident and I should have made it clear in the comments.

`ScanToEndOfLine`, whenever possible, returns a span that points to _the original bytes_, not a copy of them. This is because it's likely that we'll end up copying or appending the bytes to another buffer further down the line, so there's no point in allocating a new temporary buffer unnecessarily. For instance, if the content of the buffer looks like this—

```
[0] 'a'
[1] 'b'
[2] \n
[3] 'c'
[4] 'd'
[5] \e
```

—the first call to the method will return a span that points to bytes 0 through 1, and the second will return a span that points to bytes 3 through 4, in the buffer itself. So, we don't want to shift the data over immediately after we scan the first line, because that would overwrite the bytes that the caller is about to be looking at.

The case where `_partialLine` exists is different, because there we have already allocated a separate buffer— we had to, because the data wouldn't fit in the main fixed-size one— and so we can safely return a reference to that data while relocating the bytes in the main buffer.

Reply:
Oh right, I had known those were just views but I wasn't thinking about it when reading this code. Thanks for the explanation, that certainly explains why we can do the `BlockCopy` in the first case but not the second.